proto/prost/private/prost.bzl

```
@@ -0,0 +1,479 @@
+    additional_args.add_all(prost_toolchain.prost_opts, format_each = "--prost_opt=%s")
+
+    if is_tonic:
+        if not prost_toolchain.tonic_plugin:
+            fail("Tonic plugin not configured for this toolchain")
+
+        tonic_plugin = prost_toolchain.tonic_plugin[DefaultInfo].files_to_run
+        additional_args.add(prost_toolchain.tonic_plugin_flag % tonic_plugin.executable.path)
+        additional_args.add("--tonic_opt=no_include")
+        additional_args.add("--tonic_opt=compile_well_known_types")
+        additional_args.add("--is_tonic")
+        additional_args.add_all(prost_toolchain.tonic_opts, format_each = "--tonic_opt=%s")
+        tools = depset([tonic_plugin.executable], transitive = [tools])
+
+    if rustfmt_toolchain:
+        additional_args.add("--rustfmt={}".format(rustfmt_toolchain.rustfmt.path))
+        tools = depset(transitive = [tools, rustfmt_toolchain.all_files])
+
+    additional_inputs = depset([deps_info_file, proto_info.direct_descriptor_set] + [dep[provider].package_info for dep in deps])
+
+    proto_common.compile(
+        actions = ctx.actions,
+        proto_info = proto_info,
+        additional_tools = tools.to_list(),
+        additional_inputs = additional_inputs,
+        additional_args = additional_args,
+        generated_files = [lib_rs, package_info_file],
+        proto_lang_toolchain_info = _create_proto_lang_toolchain(ctx, prost_toolchain),
+        plugin_output = ctx.bin_dir.path,
+    )
+
+    return lib_rs, package_info_file
+
+def _get_crate_info(providers):
+    """Finds the CrateInfo provider in the list of providers."""
+    for provider in providers:
+        if hasattr(provider, "name"):
+            return provider
+    fail("Couldn't find a CrateInfo in the list of providers")
+
+def _get_dep_info(providers):
+    """Finds the DepInfo provider in the list of providers."""
+    for provider in providers:
+        if hasattr(provider, "direct_crates"):
+            return provider
+    fail("Couldn't find a DepInfo in the list of providers")
+
+def _get_cc_info(providers):
+    """Finds the CcInfo provider in the list of providers."""
+    for provider in providers:
+        if hasattr(provider, "linking_context"):
+            return provider
+    fail("Couldn't find a CcInfo in the list of providers")
+
+def _compile_rust(ctx, attr, crate_name, src, deps, edition):
+    """Compiles a Rust source file.
+
+    Args:
+      ctx (RuleContext): The rule context.
+      attr (Attrs): The current rule's attributes (`ctx.attr` for rules, `ctx.rule.attr` for aspects)
+      crate_name (str): The crate module name to use.
+      src (File): The crate root source file to be compiled.
+      deps (List of DepVariantInfo): A list of dependencies needed.
+      edition (str): The Rust edition to use.
+
+    Returns:
+      A DepVariantInfo provider.
+    """
+    toolchain = ctx.toolchains["@rules_rust//rust:toolchain_type"]
+    output_hash = repr(hash(src.path))
```

Maybe this should be different based on `is_tonic`?
```suggestion
    output_hash = repr(hash(src.path + (".tonic" if is_tonic else ".prost"))
```